File: f90code/CHP11/cone.f90
!
! Numerical Mathematics and Computing, Third Edition
! Ward Cheney & David Kincaid
! Brooks/Cole Publ. Co.
! ISBN 0-534-20112-1
! Copyright (c) 1994.  All rights reserved.
! For educational use with the Cheney-Kincaid textbook.
! Absolutely no warranty implied or expressed.
!
! Section 11.2
!
! File: cone.f90
!
! Ice cream cone example

program cone
      integer, parameter :: n = 5000, iprt = 1000
      real, dimension(n,3) :: r
      integer ::  i,m
      real :: x, y, z, vol
      m = 0
      call random_number(r)
      do i = 1,n
         x = 2.0*r(i,1) - 1.0 
         y = 2.0*r(i,2) - 1.0 
         z = 2.0*r(i,3)
         if( x*x + y*y <= z*z .and. x*x + y*y <= z*(2.0 - z) )  m = m + 1
         if( mod(i,iprt) == 0 )  then 
            vol = 8.0*real(m)/real(i) 
            print *,i,vol 
         end if    
      end do
end program cone
  

File: c++code/CHP11/cone.c
/****************************************************************/
/* Module : cone.c					        */
/* Section: 11.2					        */
/* Cheney-Kincaid, Numerical Mathematics and Computing, 4th ed, */
/* Brooks/Cole Publ. Co.                                        */
/* ISBN 0-534-20112-1                                           */
/* Copyright (c) 1999.  All rights reserved.                    */
/* For educational use with the Cheney-Kincaid textbook.        */
/* Absolutely no warranty implied or expressed.                 */   
/* 								*/
/* Programmer: Hidajaty Thajeb					*/
/* Description: Ice cream cone example                          */
/* Comment: The results might be slightly different from the    */
/*          book answers due to the different random values	*/
/*								*/
/****************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <math.h>

int main()
{
  const int MAXNUM = 2147483647,
            n = 5000,
            iprt = 1000;
  double r[n+1][3+1];
  int i, j, m;
  double vol, x, y, z;

  m = 0;
  for (i = 1; i <= n; i++)
    for (j = 1; j <= 3; j++)
      r[i][j] = (double)rand()/(double)MAXNUM;
  for (i = 1; i <= n; i++)
    {
      x = 2*r[i][1] - 1;
      y = 2*r[i][2] - 1;
      z = 2*r[i][3];
      if ( (x*x + y*y <= z*z) && (x*x + y*y <= z*(2 - z)))
	m++;
      if ( (i%iprt) == 0)
	{
	  vol = 8*(double)m/(double)i;
	  printf("i = %d, vol = %.1f\n", i, vol);
	}        	   
    }   
  return 0;
}
